File: frontend/src/services/aiService.js
import useAIStore from "../store/useAIStore";

const API_BASE = import.meta.env.VITE_API_BASE || "/api";

function getClientFallback(action, text) {
  const rawText = (text || "").trim();
  const pLower = (action || "").toLowerCase() + " " + rawText.toLowerCase();

  if (pLower.includes("birthday")) {
    return (
      "🎉 **Celebrating Special Milestones: The Ultimate Birthday Guide**\n\n" +
      "Birthdays are more than just another date on the calendar—they are a time to pause, reflect, and celebrate the incredible journey of life with the people who matter most.\n\n" +
      "### 🌟 1. Crafting Personal & Heartfelt Messages\n" +
      "The best birthday greetings are **personal, warm, and honest**. A meaningful message highlights shared memories, expresses genuine appreciation, and shares inspiring wishes for the year ahead.\n\n" +
      "### 🎁 2. Creating Unforgettable Moments\n" +
      "- **Surprise Elements:** Small, thought-out gestures leave lasting impressions.\n" +
      "- **Shared Experiences:** Quality time spent together often outshines material gifts.\n" +
      "- **Gratitude Reflections:** Taking a moment to appreciate growth and milestones over the past year.\n\n" +
      "### 🚀 3. Looking Forward to the Year Ahead\n" +
      "Every new year brings fresh opportunities, bigger dreams, and exciting adventures. Here's to making every single day count!"
    );
  } else if (pLower.includes("outline")) {
    const topic = rawText.length > 0 && rawText.length < 60 ? rawText : "Modern Content Strategy";
    return (
      `📌 **Blog Outline: ${topic.toUpperCase()}**\n\n` +
      "### 1. Executive Summary & Overview\n" +
      `- Why ${topic} is essential for digital creators today.\n` +
      "- Key takeaways and strategic insights.\n\n" +
      "### 2. Core Concepts & Fundamentals\n" +
      "- Foundational principles and framework breakdown.\n" +
      "- Practical industry examples.\n\n" +
      "### 3. Step-by-Step Implementation\n" +
      "- Actionable strategies to execute immediately.\n" +
      "- Common pitfalls to avoid.\n\n" +
      "### 4. Conclusion & Key Takeaways\n" +
      "- Summary of essential points.\n" +
      "- Recommended next steps for readers."
    );
  } else if (pLower.includes("headline") || pLower.includes("title")) {
    const subject = rawText.length > 0 && rawText.length < 40 ? rawText : "Content Creation";
    return (
      `🚀 **5 Catchy SEO Headlines for: ${subject}**\n\n` +
      `1. The Complete 2026 Guide to Mastering ${subject}\n` +
      `2. 5 Proven Strategies to Transform Your ${subject} Today\n` +
      `3. Why Top Creators Are Rethinking ${subject} in 2026\n` +
      `4. How to Scale Your ${subject} 10x Faster\n` +
      `5. The Secret Blueprint for High-Performing ${subject}`
    );
  } else if (pLower.includes("seo") || pLower.includes("meta")) {
    const topic = rawText.length > 0 && rawText.length < 40 ? rawText : "Digital Publishing";
    return (
      "🔍 **SEO Metadata Package**\n\n" +
      `**Meta Title:** Ultimate Guide to ${topic} | SmartBlog Studio\n` +
      `**Meta Description:** Discover actionable insights, expert tips, and proven strategies for ${topic} to boost audience engagement.\n` +
      `**Target Keywords:** #${topic.replace(/\s+/g, '')} #Blogging #ContentStrategy #AITools #SEO`
    );
  } else if (pLower.includes("grammar") || pLower.includes("polish")) {
    return `✨ **Polished Version:**\n\n${rawText || "Modern content creation requires high-quality writing, clear structure, and seamless publishing tools to engage audiences effectively."}`;
  } else if (pLower.includes("tone") || pLower.includes("casual") || pLower.includes("punchy")) {
    return `⚡ **Rewritten Tone Version:**\n\n${rawText ? `"${rawText}" — Re-phrased with enhanced impact, bold clarity, and dynamic sentence flow for digital readers.` : "Leveraging smart design and automated AI assistance empowers creators to produce impactful blog posts with effortless precision."}`;
  } else {
    const topicPhrase = rawText.length > 0 ? rawText : "modern digital productivity";
    return (
      `🌟 **Expanded Content: ${topicPhrase.toUpperCase()}**\n\n` +
      `Focusing on **${topicPhrase}** brings depth, clarity, and authority to your blog post. ` +
      `When creators emphasize ideas that are personal, engaging, and well-structured, readers connect far more deeply with the message.\n\n` +
      "### Key Dimensions to Consider:\n" +
      `- **Authentic Phrasing:** Framing "${topicPhrase}" with precision ensures your core message resonates effortlessly.\n` +
      "- **Audience Engagement:** Well-crafted paragraphs encourage readers to stay invested and take action.\n" +
      "- **Structural Flow:** Connecting main ideas logically creates a seamless reading experience.\n\n" +
      `By expanding on *${topicPhrase}*, your post moves beyond basic statements into an engaging narrative that inspires your audience!`
    );
  }
}

export const streamAI = async (text, action) => {
  const { setGenerating, appendResult, clearResult, setError } =
    useAIStore.getState();

  try {
    clearResult();
    setGenerating(true);

    let token = localStorage.getItem("token");
    if (!token) {
      token = "mock-dev-jwt-token";
      localStorage.setItem("token", token);
    }

    const controller = new AbortController();
    useAIStore.getState().setController(controller);

    let streamSuccess = false;

    // Layer 1: Streaming API endpoint
    try {
      const response = await fetch(`${API_BASE}/ai/stream`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({ text: text || "blog content", action: action || "expand" }),
        signal: controller.signal,
      });

      if (response.ok && response.body) {
        const reader = response.body.getReader();
        const decoder = new TextDecoder("utf-8");

        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          const chunk = decoder.decode(value, { stream: true });
          if (chunk) {
            appendResult(chunk);
            streamSuccess = true;
          }
        }
      }
    } catch (e) {
      if (e.name === "AbortError") {
        setGenerating(false);
        return;
      }
      console.warn("Stream fetch fallback:", e);
    }

    if (streamSuccess) {
      setGenerating(false);
      return;
    }

    // Layer 2: Non-streaming API endpoint
    try {
      const genRes = await fetch(`${API_BASE}/ai/generate`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({ text: text || "blog content", action: action || "expand" }),
        signal: controller.signal,
      });

      if (genRes.ok) {
        const data = await genRes.json();
        if (data.result) {
          appendResult(data.result);
          setGenerating(false);
          return;
        }
      }
    } catch (e) {
      if (e.name === "AbortError") {
        setGenerating(false);
        return;
      }
      console.warn("Generate fetch fallback:", e);
    }

    // Layer 3: Instant client-side fallback
    const fallbackText = getClientFallback(action, text);
    appendResult(fallbackText);
    setGenerating(false);

  } catch (err) {
    if (err.name === 'AbortError') {
      setGenerating(false);
      return;
    }
    const fallbackText = getClientFallback(action, text);
    appendResult(fallbackText);
    setGenerating(false);
  }
};

export const streamAIService = streamAI;
export default streamAI;

File: backend/app/services/ai_service.py
import os
from dotenv import load_dotenv
from app.database import db
from datetime import datetime

load_dotenv()

GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-2.0-flash")

client = None
if GEMINI_API_KEY:
    try:
        from google import genai
        client = genai.Client(api_key=GEMINI_API_KEY)
    except Exception as e:
        print("[AI] GenAI Client Init Exception:", e)

ai_collection = db["ai_usage"]


async def log_ai_usage(user_id: str, action: str, input_length: int):
    """Log AI usage to MongoDB — silently skips if collection unavailable."""
    try:
        ai_collection.insert_one({
            "user_id": user_id,
            "action": action,
            "input_length": input_length,
            "timestamp": datetime.utcnow()
        })
    except Exception:
        pass  # MockDB or unavailable — continue silently


async def build_prompt(text: str, action: str) -> str:
    """Build the full prompt string for a given action type."""
    if action == "summary":
        return f"Summarize the following text clearly and concisely:\n\n{text}"
    elif action == "grammar":
        return f"Fix all grammar, spelling, and punctuation errors in the following text:\n\n{text}"
    elif action == "expand":
        return f"Elaborate and expand on the following text with engaging blog content and vivid details:\n\n{text}"
    elif action == "outline":
        return f"Generate a comprehensive, structured blog post outline with section headings and subpoints for:\n\n{text}"
    elif action == "headline":
        return f"Generate 5 catchy, high-converting, SEO-optimized blog headlines for:\n\n{text}"
    elif action == "seo_meta":
        return f"Generate an SEO Meta Title, Meta Description (max 160 chars), and 5 targeted keywords for:\n\n{text}"
    elif action == "tone_casual":
        return f"Rewrite the following in a warm, conversational, and relatable tone:\n\n{text}"
    elif action == "tone_professional":
        return f"Rewrite the following in a polished, authoritative, professional business tone:\n\n{text}"
    elif action == "tone_punchy":
        return f"Rewrite the following in a bold, fast-paced, punchy, persuasive tone:\n\n{text}"
    return f"Based on the following input, write a compelling, well-structured blog article:\n\n{text}"


def generate_fallback_content(prompt: str) -> str:
    """Dynamic fallback generator when Gemini API is unavailable."""
    p = prompt.lower()
    raw = prompt.split("\n\n")[-1].strip() if "\n\n" in prompt else prompt.strip()
    topic = raw if len(raw) < 80 else "Modern Content Strategy"

    if "birthday" in p:
        return (
            "🎉 **Celebrating Special Milestones: The Ultimate Birthday Guide**\n\n"
            "Birthdays are more than just another date — they are a time to pause and celebrate life.\n\n"
            "### 🌟 1. Crafting Heartfelt Messages\nThe best birthday greetings are personal, warm, and honest.\n\n"
            "### 🎁 2. Creating Unforgettable Moments\n- Small gestures leave lasting impressions.\n- Quality time together outshines material gifts.\n\n"
            "### 🚀 3. Looking Forward\nEvery new year brings bigger dreams. Here's to making every day count!"
        )
    elif "outline" in p:
        return (
            f"📌 **Blog Outline: {topic.capitalize()}**\n\n"
            f"### 1. Introduction\n- Why {topic} matters for modern creators.\n\n"
            "### 2. Core Concepts\n- Key principles and real-world applications.\n\n"
            "### 3. Step-by-Step Guide\n- Actionable strategies to implement today.\n\n"
            "### 4. Conclusion\n- Key takeaways and recommended next steps."
        )
    elif "headline" in p:
        return (
            f"🚀 **5 Catchy Headlines for: {topic.capitalize()}**\n\n"
            f"1. The Complete 2026 Guide to Mastering {topic.capitalize()}\n"
            f"2. 5 Proven Strategies to Transform Your {topic.capitalize()} Today\n"
            f"3. Why Top Creators Are Rethinking {topic.capitalize()} in 2026\n"
            f"4. How to Scale Your {topic.capitalize()} Results 10x Faster\n"
            f"5. The Secret Blueprint for High-Performing {topic.capitalize()}"
        )
    elif "seo_meta" in p or "meta" in p:
        return (
            f"🔍 **SEO Metadata**\n\n"
            f"**Meta Title:** Ultimate Guide to {topic.capitalize()} | Smart Blog Editor\n"
            f"**Meta Description:** Discover actionable insights and proven strategies for {topic} to boost audience engagement.\n"
            f"**Target Keywords:** #{topic.replace(' ', '')} #Blogging #ContentStrategy #SEO"
        )
    elif "grammar" in p or "polish" in p or "fix" in p:
        return f"✨ **Polished Version:**\n\n{raw}\n\n*(Refined for clarity and professional tone.)*"
    elif "casual" in p:
        return f"💬 **Casual Version:**\n\nHey! Here's a friendly take:\n\n{raw}\n\nKeep it real and relatable!"
    elif "punchy" in p:
        return f"⚡ **Punchy Version:**\n\n**{raw}**\n\nFast. Bold. Impactful."
    elif "professional" in p:
        return f"🏢 **Professional Version:**\n\n{raw}\n\n*(Rewritten with authoritative tone and executive-level clarity.)*"
    else:
        return (
            f"🌟 **Expanded Content: {topic.capitalize()}**\n\n"
            f"When it comes to **{topic}**, depth and authenticity make all the difference.\n\n"
            "### Key Dimensions to Explore:\n"
            f"- **Clarity:** Frame your ideas around *{topic}* with precision.\n"
            "- **Engagement:** Well-crafted paragraphs keep readers invested.\n"
            "- **Flow:** Connect ideas logically for a seamless reading experience.\n\n"
            f"By expanding on *{topic}*, your content becomes an inspiring narrative!"
        )


def stream_ai_response(prompt: str):
    """Synchronous generator yielding bytes for StreamingResponse."""
    if client:
        try:
            response = client.models.generate_content_stream(
                model=GEMINI_MODEL,
                contents=prompt,
            )
            for chunk in response:
                try:
                    text = getattr(chunk, "text", None)
                    if text:
                        yield text.encode("utf-8")
                except Exception:
                    continue
            return
        except Exception as e:
            print("[AI] Gemini API call failed, using fallback:", e)

    # Client-side fallback
    yield generate_fallback_content(prompt).encode("utf-8")
